javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -0,0 +1,164 @@
+ * Modifying the tree is done similarly, but along the way the path is copied, returning a new root every time.
+ * Append inserts in the last leaf, or if the tree is full from the right, it adds another layer on top of it (the old root will be the first of the new one).
+ * Prepend is done similarly, but an offset is needed, because adding a new top node (where the current root would be the last node of the new root)
+ * shifts the indices by half of the current tree's full size. The `offset` shifts them back to the correct index.
+ *
+ * @author Pap Lőrinc
+ * @since 2.1.0
+ */
+final class BitMappedTrie<T> implements Serializable {
+    static int BRANCHING_BASE = 5;
+    static int branchingFactor()                   { return 1 << BRANCHING_BASE; }
+    static int firstDigit(int num, int depthShift) { return num >> depthShift; }
+    static int digit(int num, int depthShift)      { return lastDigit(firstDigit(num, depthShift)); }
+    static int lastDigit(int num)                  { return num & (-1 >>> -BRANCHING_BASE); }
+
+    private static final long serialVersionUID = 1L;
+
+    private static final BitMappedTrie<?> EMPTY = new BitMappedTrie<>(Arrays.empty(), 0, 0, 0);
+    @SuppressWarnings("unchecked")
+    static <T> BitMappedTrie<T> empty() { return (BitMappedTrie<T>) EMPTY; }
+
+    private final Object[] array;
+    private final int offset, length;
+    private final int depthShift;
+
+    BitMappedTrie(Object[] array, int offset, int length, int depthShift) {
+        this.array = array;
+        this.offset = offset;
+        this.length = length;
+        this.depthShift = depthShift;
+
+        assert length() <= treeSize(branchingFactor(), depthShift);
+    }
+
+    private static int treeSize(int branchCount, int depthShift) {
+        final int fullBranchSize = 1 << depthShift;
+        return branchCount * fullBranchSize;
+    }
+
+    BitMappedTrie<T> prepend(T leading) {
+        final int newSize = length() + 1;
+        if (length() == 0) {
+            return new BitMappedTrie<>(asArray(leading), offset, newSize, depthShift);
+        } else {
+            Object[] array = this.array;
+            int shift = depthShift, offset = this.offset;
+            if (isFullLeft()) {
+                array = copyUpdate(Arrays.empty(), branchingFactor() - 1, array);
+                shift += BRANCHING_BASE;
+                offset = treeSize(branchingFactor() - 1, shift);
+            }
+
+            offset -= 1;
+            array = modifyLeaf(array, shift, offset, COPY_LEAF, updateLeafWith(leading));
+            return new BitMappedTrie<>(array, offset, newSize, shift);
+        }
+    }
+    private boolean isFullLeft() { return offset == 0; }
+
+    BitMappedTrie<T> append(T trailing) {
+        final int newSize = length() + 1;
+        if (length() == 0) {
+            return new BitMappedTrie<>(asArray(trailing), offset, newSize, depthShift);
```

Comment:
Do you mean treating all non-full arrays (i.e. first and last, the rest should be full) as semi-mutable arrays (i.e. masked out parts to be modifiable)?
In simple cases it could probably be done, but not if we allow valid `null` values too (If I `prepend`, `drop` and `prepend`, while keeping all the intermediary states, the second prepend could only override values, if the value was `null`, but that might have been inserted by the first `prepend`), i.e.

* `of(1)` -> (1)
* `prepend(null)` -> (null, 1)
* `drop(1)` -> null, (1)
* `prepend(1)` -> (1, 0)